Add a psalm type for field mapping

laryjulien opened the pull request:
Field mapping have different definitions
in property definition and method return.
As suggested in issue and to avoid further desynchronization,
a psalm type has been created.
Fixes #9193

greg0ire commented:
@laryjulien thanks! Both Psalm and Phpstan now detect 7 issues with keys that are accessed without checking first if they might be null… for each of them, I see 3 possibilities:
- we `assert` that it is not null;
- we check that it is not null, throw a custom exception if it is;
- we ignore the error in the baseline of both tools.

I'm not sure what's best, but I think it mostly depends on whether the issue could actually happen or not.

azjezz commented on lib/Doctrine/ORM/Mapping/Driver/AnnotationDriver.php:
```diff
@@ -712,18 +715,7 @@ private function joinColumnToArray(Mapping\JoinColumn $joinColumn): array
-     *                   columnName?: string,
-     *                   columnDefinition?: string
-     *               }
+     * @psalm-return FieldMapping
```
this is a BC break IMHO, previous type states that `fieldName`, `type`, `scale`, `length`, `unique`, `nullable`, and `precision` are always available, while the new type states that they might not exist in the returned array.

greg0ire replied:
Oh right, didn't catch that, why did you change this @laryjulien ?

laryjulien replied:
I have changed my PR to only use FieldMapping type in the initial context of issue.
I suggest to open an other PR to use this type in a wider context. Indeed, there are small variations on field mapping definition which appear to be more complex than I thought at first glance.

azjezz commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -1534,7 +1532,7 @@ private function validateAndCompleteTypedAssociationMapping(array $mapping): arr
      * Validates & completes the given field mapping.
      *
-     * @psalm-param array<string, mixed> $mapping The field mapping to validate & complete.
+     * @psalm-param FieldMapping $mapping The field mapping to validate & complete.
```
This is actually the BC break, as the other method is private.

laryjulien committed:
Add a psalm type for field mapping

Field mapping have different definitions
in property definition and method return.
As suggested in issue and to avoid further desynchronization,
a psalm type has been created.
Fixes #9193

derrabus commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -70,6 +70,26 @@
+ *      requireSQLConversion?: bool,
+ *      declared?: class-string,
+ *      declaredField?: string,
+ *      options: array<mixed>
```
```suggestion
 *      options?: array<string, mixed>
```

laryjulien replied:
Updated

derrabus reviewed (commented):
Thank you very much for working on this issue. Just today I performed an ORM update on a codebase and I suddenly had lots of false-positive Psalm errors because of those inconsitently documented array shapes. Looking forward to seeing this merged. 👍🏻

greg0ire commented:
Thanks @laryjulien !
